File: python/message.py
HANDSHAKE = 10
CHOKE = 0
UNCHOKE = 1
INTERESTED = 2
NOT_INTERESTED = 3
HAVE = 4
BITFIELD = 5
REQUEST = 6
PIECE = 7


class Message:

    def __init__(self, length, type_, payload):
        self.length = length
        self.type = type_
        self.payload = payload

    def get_type(self):
        if self.type == HANDSHAKE:
            return 'handshake'
        elif self.type == CHOKE:
            return 'choke'
        elif self.type == UNCHOKE:
            return 'unchoke'
        elif self.type == INTERESTED:
            return 'interested'
        elif self.type == NOT_INTERESTED:
            return 'not_interested'
        elif self.type == HAVE:
            return 'have'
        elif self.type == BITFIELD:
            return 'bitfield'
        elif self.type == REQUEST:
            return 'request'
        elif self.type == PIECE:
            return 'piece'
        else:
            return 'INVALID'

    def to_bytes(self):
        return self.length.to_bytes(4, 'big') + self.type.to_bytes(1, 'big') + self.payload


def handshake(peer_id):
    return bytes('P2PFILESHARINGPROJ', 'ascii') + b'\0\0\0\0\0\0\0\0\0\0' + peer_id.to_bytes(4, 'big')


def choke():
    return Message(1, CHOKE, b'').to_bytes()


def unchoke():
    return Message(1, UNCHOKE, b'').to_bytes()


def interested():
    return Message(1, INTERESTED, b'').to_bytes()


def not_interested():
    return Message(1, NOT_INTERESTED, b'').to_bytes()


def have(index):
    return Message(1 + 4, HAVE, index.to_bytes(4, 'big')).to_bytes()


def bitfield(arr):
    payload = arr.tobytes()
    return Message(1 + len(payload), BITFIELD, payload).to_bytes()


def request(index):
    return Message(1 + 4, REQUEST, index.to_bytes(4, 'big')).to_bytes()


def piece(index, file_piece):
    payload = index.to_bytes(4, 'big') + file_piece
    return Message(1 + len(payload), PIECE, payload).to_bytes()


def parse(msg):
    if msg[:18] == bytes('P2PFILESHARINGPROJ', 'ascii'):
        return Message(1 + 4, HANDSHAKE, msg[28:32])
    else:
        length = int.from_bytes(msg[:4], 'big')
        return Message(length, msg[4], msg[5:4 + length])

File: java2/Message.java
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.BitSet;

public class Message {
    public static String P2PFILESHARINGPROJ = "P2PFILESHARINGPROJ";

    public final int length;
    public final MessageType type;
    public final byte[] payload;

    public Message(int length, MessageType type, byte[] payload) {
        this.length = length;
        this.type = type;
        this.payload = payload;
    }

    public byte[] toBytes() {
        return concat(intToBytes(this.length), new byte[] {(byte) this.type.code}, payload);
    }

    private static byte[] concat(byte[] first, byte[]... rest) {
        int totalLength = first.length;
        for (byte[] array : rest) {
            totalLength += array.length;
        }
        byte[] result = Arrays.copyOf(first, totalLength);
        int offset = first.length;
        for (byte[] array : rest) {
            System.arraycopy(array, 0, result, offset, array.length);
            offset += array.length;
        }
        return result;
    }

    public static byte[] handshake(int peerId) {
        return concat(P2PFILESHARINGPROJ.getBytes(), new byte[10], intToBytes(peerId));
    }

    public static byte[] choke()  {
        return new Message(1, MessageType.CHOKE, new byte[] {}).toBytes();
    }

    public static byte[] unchoke() {
        return new Message(1, MessageType.UNCHOKE, new byte[] {}).toBytes();
    }

    public static byte[] interested()  {
        return new Message(1, MessageType.INTERESTED, new byte[] {}).toBytes();
    }

    public static byte[] notinterested()  {
        return new Message(1, MessageType.NOT_INTERESTED, new byte[] {}).toBytes();
    }

    public static byte[] have(int index)  {
        return new Message(5, MessageType.HAVE, intToBytes(index)).toBytes();
    }

    public static byte[] bitfield(BitSet bitarray) {
        return new Message(5 + (bitarray.length() + 7)/8, MessageType.BITFIELD, bitarray.toByteArray()).toBytes();
    }

    public static byte[] request(int index) {
        return new Message(5, MessageType.REQUEST, intToBytes(index)).toBytes();
    }

    public static byte[] piece(int index, byte[] filePiece) {
        return new Message(5, MessageType.PIECE, concat(intToBytes(index), filePiece)).toBytes();
    }

    public static byte[] intToBytes(int n) {
        ByteBuffer byteBuffer = ByteBuffer.allocate(4);
        byteBuffer.putInt(n);
        return byteBuffer.array();
    }

    public static Message parse(byte[] msg) {
        if (new String(Arrays.copyOfRange(msg, 0, 18)).equals(P2PFILESHARINGPROJ)) {
            return new Message(1, MessageType.HANDSHAKE, Arrays.copyOfRange(msg, 28, 32));
        } else {
            return new Message(bytesToInt(Arrays.copyOfRange(msg, 0, 4)),
                               MessageType.valueOf(msg[4]),
                               Arrays.copyOfRange(msg, 5, msg.length));
        }
    }

    private static int bytesToInt(byte[] b) {
        return ByteBuffer.wrap(b).getInt();
    }
}
